### PR TITLE
Issue/fix gradle filter

### LINKED ISSUE #3247: Kotest should respect test `filter { }` defined in gradle.build.kts
Kotest does not include filters defined within build files

i.e. something like this doesn't get properly picked up.
```kotlin
tasks {
  useJUnitPlatform()
  withType<Test> {
    when(name) {
      "configIT" -> {
        filter {
          includeTestsMatching("*ConfigurationIT")
        }
      }
    }
  }
}
```

I put a debug point in there and realized that Gradle's `ClassMethodNameFilter.matcher.commandLineIncludePatterns` returned an empty list, whereas `ClassMethodNameFilter.matcher.buildScriptIncludePatterns`  returned the supplied pattern.

image

**Which version of Kotest are you using**
5.5.1

### LINKED ISSUE #3250: Gradle test regex are incorrectly applied
# Context
`./gradlew test --tests A*` skipped all the tests in 5.5.1 - it used to work in 5.4.x.  The new modifications to kotest's Gradle Filter may have broken this functionality.

```
...AccountDetailsTest > json serialization/deserialization should round trip SKIPPED

AccountDetailsTest STANDARD_OUT

    ~~~ AccountDetailsTest ~~~
      + json serialization/deserialization should round trip: TestResult=Ignored(reason=json serialization/deserialization should round trip is excluded by test filter(s))
```

This is because Kotest's test filter was unable to translate that to correct path / classname / package name. 

# Potential Fix
Gradle supplies a well-formed regex to the engine that we can leverage to construct a regex object:
- `A*` becomes `\QA\E.*`
- `A*Test` becomes `\QA\E.*\QTest\E`
- `io.*.A*Test` becomes `\Qio.\E.*\Q.A\E.*\QTest\E`
- `io.*.A*Test.AccountDetails.fromString.*` becomes `\Qio.\E.*\Q.A\E.*\QTest.AccountDetails.fromString.\E.*`
- `io.*.A*Test.some test context.test name` becomes `\Qio.\E.*\Q.A\E.*\QTest.some test context.test name\E`

This means we can match the tests natively based on the complete path of tests.

**Which version of Kotest are you using**
5.5.1

### PR DESCRIPTION
**Pull request by myuwono**
# In this PR
- reimplement `GradleClassMethodRegexTestFilter` to translate patterns supplied by gradle to the proper test filters (fixes #3250).
  - update `GradleClassMethodRegexTestFilterTest.kt` tests to reflect the pattern supplied by latest gradle (gradle 7.5.1). 
  - update tests to using data-driven tests to make it easier to expand test patterns as well as pinpoint exactly the failing patterns in CI.
- include introspections of filters defined in buildScripts (fixes #3247)

## Patterns that works
tested against gradle 7.5.1, using both filters defined in `build.gradle.kts` and command line:
- `gradle test --tests "SomeTest"`
- `gradle test --tests "*Test"`
- `gradle test --tests "io.package.*"`
- `gradle test --tests "io.package"`
- `gradle test --tests "io.package.SomeTest"`
- `gradle test --tests "io.package.SomeTest.first level context*"`
- `gradle test --tests "io.package.SomeTest.*"`
- `gradle test --tests "io.*.SomeTest"`
- `gradle test --tests "io.*.SomeTest.*"`
- `gradle test --tests "SomeTest.first level context*"`
- `gradle test --tests "*.first level context*"`

## Pattern that does NOT work
In this PR I decided to omit the support for nested context / test matching. It's arguably quite involved so I'd reserve it for future contribution. Kotest support lazy test registration within nested context. Gradle test filter does not natively work nicely with kotest. In order to make it work we need to think of a way to recursively apply partial context-search as we dive deeper into the contexts. I.e. These patterns DOES NOT work yet:
- `gradle test --tests "SomeTest.first level context -- should assert this test description"`
- `gradle test --tests "SomeTest.first level context -- nested context -- *"`
- `gradle test --tests "SomeTest.first level context -- nested context -- yet another nested context -- *"`

# Mechanism
Gradle supplies a pattern string which corresponds to a well-formed regex object. This can be directly usable for kotest.
- `A*` becomes `\QA\E.*`
- `A*Test` becomes `\QA\E.*\QTest\E`
- `io.*.A*Test` becomes `\Qio.\E.*\Q.A\E.*\QTest\E`
- `io.*.A*Test.AccountDetails*` becomes `\Qio.\E.*\Q.A\E.*\QTest.AccountDetails\E.*`
- `io.*.A*Test.some test context*` becomes `\Qio.\E.*\Q.A\E.*\QTest.some test context\E.*`

Within the code we get this pattern string and directly translate them into a regex object to be used against the path / classname / packagename as appropriate.

### CODE REVIEW

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 1]
```diff
@@ -13,70 +14,85 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
       logger.log { Pair(descriptor.toString(), "Testing against $patterns") }
       return when {
          patterns.isEmpty() -> TestFilterResult.Include
-         patterns.any { match(it, descriptor) } -> TestFilterResult.Include
```
Comment (myuwono) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:16]: This was wrongly combining the filters. For instance if we have two in the list `*Test` defined in build.kts and `CacheTest` in command line, you'd want to omit all other tests except for `CacheTest`. This flip from `any` to `all` fixes it.

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradlePostDiscoveryFilterExtractor.kt [hunk 2]
```diff
@@ -52,6 +59,12 @@ object GradlePostDiscoveryFilterExtractor {
       return field.get(obj) as List<Any>
    }
 
+   private fun buildScriptIncludePatterns(obj: Any): List<Any> {
+      val field = obj::class.java.getDeclaredField("buildScriptIncludePatterns")
```
Comment (myuwono) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradlePostDiscoveryFilterExtractor.kt:63]: added support for filters defined in build script

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt [hunk 3]
```diff
@@ -13,70 +14,85 @@ class GradleClassMethodRegexTestFilter(private val patterns: List<String>) : Tes
...
+      val isFullPathMatched by lazy { path.matches(regexPattern) } // io.*.SomeTest
+      val isFullPathDotMatched by lazy { "$path.".matches(regexPattern) } // io.*. or io.*.SomeTest.*
+
+      // if there's no uppercase in the supplied pattern, activate trigger relaxed matching
+      val doesNotContainUppercase by lazy { pattern.replace("\\Q", "").replace("\\E", "").all { !it.isUpperCase() } }
+
+      val isPackageMatched by lazy { doesNotContainUppercase && packagePath.matches(laxRegexPattern) } // io.kotest
+      val isPackageWithDotMatched by lazy { doesNotContainUppercase && "$packagePath.".matches(laxRegexPattern) } // io.kotest.*
+
+      return isSimpleClassMatch ||
```
Comment (myuwono) [inline kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilter.kt:72]: @sksamuel @Kantis this is the change

File: kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradlePostDiscoveryFilterExtractor.kt (hunk 2 above)
Comment (Kantis) [reply kotest-runner/kotest-runner-junit5/src/jvmMain/kotlin/io/kotest/runner/junit/platform/gradle/GradlePostDiscoveryFilterExtractor.kt:63]: Are there build script exclude patterns as well? Not saying we need to fix it now, but could be flagged up as a potential future contribution

File: kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilterTest.kt [hunk 4]
```diff
@@ -9,33 +9,31 @@ import io.kotest.matchers.shouldBe
...
+         listOf("\\QGradleClassMethodRegexTestFilterTest\\E"),
+         listOf(".*\\QthodRegexTestFilterTest\\E"),
+         listOf(".*\\QTest\\E"),
+         listOf("\\Qio.kotest.runner.junit.platform.gradle.GradleClassMethodRegexTestFilterTest\\E"),
+         listOf(".*\\Q.platform.gradle.GradleClassMethodRegexTestFilterTest\\E"),
+         listOf(".*\\Qorm.gradle.GradleClassMethodRegexTestFilterTest\\E")
+      ) { filters ->
+         GradleClassMethodRegexTestFilter(filters).filter(spec) shouldBe TestFilterResult.Include
+      }
+   }
```
Comment (Kantis) [inline kotest-runner/kotest-runner-junit5/src/jvmTest/kotlin/io/kotest/runner/junit/platform/gradle/GradleClassMethodRegexTestFilterTest.kt:25]: Nice :)

**Review by Kantis — APPROVED** [review]

**Pull request merged**